### ISSUE TITLE
Error occurred while processing adopter general data

### ISSUE THREAD

**Issue by Rillke** [opencast/opencast#4310]
**Describe the bug**
A clear and concise description of what you see:
*In the Opencast log of our Opencast staging instance, we see "Error occurred while processing adopter general data"*

**To Reproduce**
?

```
2022-10-10T12:04:54,404 | ERROR | (ScheduledDataCollector:222) - Error occurred while processing adopter general data.
java.lang.RuntimeException: HttpStatus: 400, HttpResponse: <!doctype html><html lang=en><title>400 Bad Request</title><h1>Bad Request</h1><p>The browser (or proxy) sent a request that this server could not understand.</p>
	at org.opencastproject.adopter.statistic.Sender.send(Sender.java:156) ~[!/:?]
	at org.opencastproject.adopter.statistic.Sender.send(Sender.java:120) ~[!/:?]
	at org.opencastproject.adopter.statistic.Sender.sendGeneralData(Sender.java:82) ~[!/:?]
	at org.opencastproject.adopter.statistic.ScheduledDataCollector.run(ScheduledDataCollector.java:220) [!/:?]
	at java.util.TimerThread.mainLoop(Timer.java:556) [?:?]
	at java.util.TimerThread.run(Timer.java:506) [?:?
```

**Environment Information**
 - Opencast Docker 12.3

**Comment by gregorydlogan** [comment]
Sigh. I'm gojng to have to implement #4254 to fix this, aren't I.

You haven't made any changes to registration url override, right?

**Comment by Rillke** [comment]
> You haven't made any changes to registration url override, right?

Not intentionally at least.

Our production system also logs this kind of error:

```
admin_1               | 2022-10-10T23:14:31,048 | ERROR | (ScheduledDataCollector:222) - Error occurred while processing adopter general data.
admin_1               | java.lang.RuntimeException: HttpStatus: 400, HttpResponse: <!doctype html><html lang=en><title>400 Bad Request</title><h1>Bad Request</h1><p>The browser (or proxy) sent a request that this server could not understand.</p>
admin_1               | 	at org.opencastproject.adopter.statistic.Sender.send(Sender.java:156) ~[!/:?]
admin_1               | 	at org.opencastproject.adopter.statistic.Sender.send(Sender.java:120) ~[!/:?]
admin_1               | 	at org.opencastproject.adopter.statistic.Sender.sendGeneralData(Sender.java:82) ~[!/:?]
admin_1               | 	at org.opencastproject.adopter.statistic.ScheduledDataCollector.run(ScheduledDataCollector.java:220) [!/:?]
admin_1               | 	at java.util.TimerThread.mainLoop(Timer.java:556) [?:?]
admin_1               | 	at java.util.TimerThread.run(Timer.java:506) [?:?]
```

Seems like the registration server gives an unexpected (400) response?

> Sigh. I'm gojng to have to implement https://github.com/opencast/opencast/issues/4254 to fix this, aren't I.

How would this help?

**Comment by gregorydlogan** [comment]
> Seems like the registration server gives an unexpected (400) response?

Yep.

> > Sigh. I'm gojng to have to implement #4254 to fix this, aren't I.
> 
> How would this help?

Well, a 400 is either you sending something garbled enough that the server is choking on it, or the server is somehow broken but still responding otherwise.  I'll take a look tomorrow, but being able to verify that you aren't sending bad data is a good first step too.

**Comment by gregorydlogan** [comment]
@Rillke if this is still hitting you, can you try applying https://github.com/opencast/opencast/pull/4320 to your staging system and then sending me the output you get?

**Comment by Rillke** [comment]
still getting this:

```
admin_1               | 2022-10-17T23:14:30,670 | ERROR | (ScheduledDataCollector:222) - Error occurred while processing adopter general data.
admin_1               | java.lang.RuntimeException: HttpStatus: 400, HttpResponse: <!doctype html><html lang=en><title>400 Bad Request</title><h1>Bad Request</h1><p>The browser (or pro
xy) sent a request that this server could not understand.</p>
admin_1               |         at org.opencastproject.adopter.statistic.Sender.send(Sender.java:156) ~[!/:?]
admin_1               |         at org.opencastproject.adopter.statistic.Sender.send(Sender.java:120) ~[!/:?]
admin_1               |         at org.opencastproject.adopter.statistic.Sender.sendGeneralData(Sender.java:82) ~[!/:?]
admin_1               |         at org.opencastproject.adopter.statistic.ScheduledDataCollector.run(ScheduledDataCollector.java:220) [!/:?]
admin_1               |         at java.util.TimerThread.mainLoop(Timer.java:556) [?:?]
admin_1               |         at java.util.TimerThread.run(Timer.java:506) [?:?]
```

so i'm going to test your pr and send you the output :)

**Comment by lkiesow** [comment]
That should match this error from the reverse proxy:
```
2022/10/17 23:14:30 [error] 389868#0: *41991 connect() failed (111: Connection refused) while connecting to upstream, client: 141.48.2.54, server: register.opencast.org, request: "POST /api/1.0/adopter HTTP/1.1", upstream: "http://[::1]:8080/api/1.0/adopter", host: "register.opencast.org"
```

**Comment by Rillke** [comment]
> connect() failed (111: Connection refused)  while connecting to upstream

Sounds like something's wrong with the registration server at this time?

**Comment by Rillke** [comment]
This is the data from our staging-system. Replaced some strings for obvious reasons. Didn't change the `hosts`.`services` though. It already contained the ellipsis.

Adopter Information:

```json
{
  "adopter_key": "333eXXXX-b86d-XXXX-9adf-ce3b6b0aXXXX",
  "organisation_name": "Martin-Luther-Universität Halle-Wittenberg",
  "department_name": "ITZ - Abt. Anwendungssysteme",
  "first_name": "Rainer",
  "last_name": "Rillke",
  "country": "DE",
  "city": "Halle (Saale)",
  "postal_code": "06120",
  "street": "Kurt-Mothes-Str.",
  "street_no": "1",
  "email": "opencast@elearning.uni--halle.de",
  "contact_me": true,
  "send_errors": true,
  "send_usage": true
}
```
Statistical Information:

```json
{
  "statistic_key": "928eXXXX-1439-XXXX-88d1-2abdfa3fXXXX",
  "adopter_key": "333eXXXX-b86d-XXXX-9adf-ce3b6b0aXXXX",
  "job_count": 3933,
  "event_count": 248,
  "series_count": 80,
  "user_count": 6,
  "ca_count": 2,
  "total_minutes": 10677,
  "tenant_count": 1,
  "hosts": [
    {
      "cores": 6,
      "max_load": 6,
      "memory": 6306136064,
      "hostname": "admin",
      "disk_space": 0,
      "services": "org.opencastproject.smil,\norg.opencastproject.uico[...]"
    },
    {
      "cores": 4,
      "max_load": 4,
      "memory": 1031798784,
      "hostname": "presentation",
      "disk_space": 0,
      "services": "org.opencastproject.engage.plugin.manager,\norg.ope[...]"
    },
    {
      "cores": 6,
      "max_load": 6,
      "memory": 6306136064,
      "hostname": "admin",
      "disk_space": 0,
      "services": "org.opencastproject.uiconfig,\norg.opencastproject.[...]"
    }
  ],
  "version": "12.3.0"
}
```

**Comment by gregorydlogan** [comment]
That looks entirely reasonable.  Unfortunately there's no logging on the backend - something that should clearly be added.  Are your node using the default profiles?  No extra services?  The only thing I can see in that output that could be causing issues is if the list of services is somehow way too long, but you'd have to have significantly altered the node profiles I would think.

**Comment by Rillke** [comment]
> but you'd have to have significantly altered the node profiles I would think.

Didn't touch them AFAIK. `connect() failed (111: Connection refused) while connecting to upstream` means that nginx can't even successfully TCP/TLS connect to the upstream registration server. Therefore, I'd say the payload sent won't really matter.

**Comment by gregorydlogan** [comment]
~I~ Lars noticed after I posted that the nginx upstream was set to `http://localhost:8080` rather than `http://127.0.0.1:8080`.  There were a number of ipv6 related errors in the nginx connection logs associated with http POST events and failures to connect.  I'm going to guess that (randomly?) it was choosing ipv6 over 4, and then failing.

I've deployed an updated config.  If this resolves your issue then it was related to that :)  In which case I'm surprised we have any registration data in the backend at all since the actual registration service was only listening on ipv4...
